### PR TITLE
add [[BucketizedCol]] Operation

### PR DESCRIPTION
**Pull request by tosky001**
## What changes were proposed in this pull request?

add BucketizedCol Operation, which represents discretized dense input.

## How was this patch tested?

unit tests

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/BucketizedCol.scala [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+object BucketizedCol {
+  def apply[T: ClassTag](
+    boundaries: Array[Double])
+    (implicit ev: TensorNumeric[T]): BucketizedCol[T]
+  = new BucketizedCol[T](
+    boundaries = boundaries
+  )
+}
```
Comment (sperlingxx) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/BucketizedCol.scala]: Need a new line here

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/BucketizedCol.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  override def updateOutput(input: Tensor[T]): Tensor[Int] = {
+
+    val rows = input.size(dim = 1)
+    val cols = input.size(dim = 2)
+    val resTensor = Tensor[Int](rows, cols)
+    require(boundaries.length >= 1,
+    "the length of boundaries must be more than or equal to 1")
+    var row = 1
+    while (row <= rows) {
```
Comment (sperlingxx) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/BucketizedCol.scala]: Maybe _DenseTensorApply.apply1_ is more reliable way than iterating _Tensor_ by hand?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/BucketizedCol.scala [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+  private val boundariesImpl = boundaries.map(ev.fromType[Double])
+  output = Tensor[Int]()
+
+  override def updateOutput(input: Tensor[T]): Tensor[Int] = {
+
+    val rows = input.size(dim = 1)
+    val cols = input.size(dim = 2)
+    val resTensor = Tensor[Int](rows, cols)
+    require(boundaries.length >= 1,
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/BucketizedCol.scala]: move this out of the method. Looks like this should be checked when constructing the class, right?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/BucketizedCol.scala [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    ev.getType() match {
+      case FloatType =>
+        resTensor.applyFun[Float](
+          input.asInstanceOf[Tensor[Float]],
+          x => math.abs(binarySearch(boundariesImpl.asInstanceOf[Array[Float]], x) + 1))
+      case DoubleType =>
+        resTensor.applyFun[Double](
+          input.asInstanceOf[Tensor[Double]],
+          x => math.abs(binarySearch(boundariesImpl.asInstanceOf[Array[Double]], x) + 1))
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/BucketizedCol.scala]: should we throw an exception for not support numeric type here?

**Comment by tosky001** [comment]
@yiheng the problems have been modified and thanks for your reviews.

**Comment by yiheng** [comment]
Jenkins test failure:
```
 java.lang.NoSuchFieldException: boundaries
      at java.lang.Class.getDeclaredField(Class.java:1961)
      at com.intel.analytics.bigdl.utils.serializer.ModuleSerializable$$anonfun$doSerializeModule$1.apply(ModuleSerializable.scala:218)
      at com.intel.analytics.bigdl.utils.serializer.ModuleSerializable$$anonfun$doSerializeModule$1.apply(ModuleSerializable.scala:208)
      at scala.collection.immutable.List.foreach(List.scala:318)
```

Please add a ```private val``` to boundaries declaration. Or scala erase the field for storage optimization...

**Comment by tosky001** [comment]
@yiheng Done and the Jenkins test successfully.

**Pull request merged**